core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -983,6 +984,7 @@ public static Set<Rule<?>> columnPruningRules(Metadata metadata)
     {
         return ImmutableSet.of(
                 new PruneAggregationColumns(),
+                new RemoveEmptyGlobalAggregation(), // aggregation can become empty after pruning its output columns
```

Comment:
In fact, the new rule is not a "column pruning rule", so it does not fit in this set. I think it should be inlined into `PruneAggregationColumns`.

Reply:
> I think it should be inlined into `PruneAggregationColumns`.

I thought about that. That depends on whether we assume that empty global aggregation can be created only as a result of `PruneAggregationColumns`

I didn't want to assume that. Should i?

Reply:
I think we can assume that and inline the rule. We might add a sanity check to make sure that no empty global aggregation survived. 

On the other hand, we've got a bunch of "cleanup" rules, including e.g. `RemoveRedundantLimit`. I think that `RemoveEmptyGlobalAggregation` is a good fit. However, those rules are used only once, in an early optimization phase, where empty global aggregations are not yet present. I wonder if we shouldn't run those rules more often, like the column pruning rules. Then we could add `RemoveEmptyGlobalAggregation` to that set.

Reply:
if we were to inline the logic into `PruneAggregationColumns`, this would mean we have a contract that the there are no empty global aggregations. Such a contract should be enforced in `AggregationNode` constructor. Will this be OK?

BTW empty global aggregation can be result of column pruning, but can also be result of initial planning:

```
SELECT 1 FROM tpch.tiny.nation GROUP BY GROUPING SETS (())
```

(for this query, during planning we indeed create `AggregationNode` with no aggregate functions)

would this call for a separate rule?

Reply:
> Such a contract should be enforced in AggregationNode constructor

I don't think we should restrict creating empty global aggregations, especially that they are legit and correct. From what I understand, efficiency is the only issue.

> would this call for a separate rule?

Yes. If both the QueryPlanner, and the pruning rule create empty global aggregations, we should have a separate rule to optimize them. However, we shouldn't add this rule to `columnPruningRules` just for convenience.

Reply:
> However, we shouldn't add this rule to `columnPruningRules` just for convenience.

Sure, i admit i was being lazy. Where would you rather see it?

Reply:
why not put if together with some pruning rules, not after them? we know this can be an effect of the pruning.

adding new optimizers isn't generally desirable

Reply:
That would imply it can only be the effect of pruning while we know that this can also be the effect of initial planning. Anyway, I can see that the pruning Optimizer is a convenient location, and we might not be able to come up with a better one.